cmd/generate-catalog/generate.go

```diff
@@ -0,0 +1,227 @@
+	// needed for digest algorithm validation
+	_ "crypto/sha256"
+
+	"encoding/base64"
+	"fmt"
+	"log"
+	"os"
+	"slices"
+
+	semver "github.com/Masterminds/semver/v3"
+	"github.com/distribution/reference"
+	"github.com/goccy/go-yaml"
+)
+
+const (
+	inputFile                       = "bundles.yaml"
+	outputFile                      = "catalog-template.yaml"
+	iconFile                        = "icon.png"
+	firstLineHeadComment            = "DO NOT EDIT. This file is generated by cmd/generate-catalog/generate.go script."
+	secondLineHeadComment           = "Any manual changes will be overwritten."
+	channelDeprecationMessage       = "This version is no longer supported. Please switch to the `stable` channel or a channel for a version that is still supported.\n"
+	bundleDeprecationMessage        = "This operator bundle version is no longer supported. Please switch to non deprecated bundle version for support.\n"
+	deprecationMessageLatestChannel = "The `latest` channel is no longer supported.  Please switch to the `stable` channel.\n"
+)
+
+func main() {
+	input, err := readInputFile()
+	if err != nil {
+		log.Fatalf("Failed to read bundle list file: %v", err)
+	}
+
+	versions := make([]*semver.Version, 0)
+	for _, img := range input.Images {
+		versions = append(versions, img.Version)
+	}
+
+	versionToImageMap, err := buildMapVersionToImage(input.Images)
+	if err != nil {
+		log.Fatalf("Failed to parse versions: %v", err)
+	}
+
+	pkg, err := generatePackageWithIcon()
+	if err != nil {
+		log.Fatalf("Failed to generate package object with icon: %v", err)
+	}
+	channels := generateChannels(versions, input.BrokenVersions)
+	deprecations := generateDeprecations(versions, input.OldestSupportedVersion)
+	bundles := generateBundles(versions, versionToImageMap)
+
+	ct := newCatalogTemplate()
+	ct.addPackage(pkg)
+	ct.addChannels(channels)
+	ct.addDeprecations(deprecations)
+	ct.addBundles(bundles)
+
+	err = ct.writeToFile()
+	if err != nil {
+		log.Fatalf("Failed to write catalog template to file: %v", err)
+	}
+
+	fmt.Printf("%s generated successfully.\n", outputFile)
+}
+
+// readInputFile unmarshals YAML file onto Input structure.
+func readInputFile() (Input, error) {
+	inputBytes, err := os.ReadFile(inputFile)
+	if err != nil {
+		return Input{}, fmt.Errorf("failed to read %s: %v", inputFile, err)
+	}
+	var input Input
+	if err := yaml.Unmarshal(inputBytes, &input); err != nil {
+		return Input{}, fmt.Errorf("failed to parse YAML: %v", err)
+	}
+
+	for i := 0; i < len(input.Images)-1; i++ {
+		version := input.Images[i].Version
+		nextVersion := input.Images[i+1].Version
+		if version.GreaterThan(nextVersion) {
+			return Input{}, fmt.Errorf("operator bundle images are not sorted in ascending order: %s > %s", version.Original(), nextVersion.Original())
+		}
+	}
+
+	return input, nil
+}
+
+// buildMapVersionToImage returns a mapping from version to BundleImage.
+func buildMapVersionToImage(images []BundleImage) (map[*semver.Version]BundleImage, error) {
```

Comment:
Renaming suggestion: how about
```suggestion
func mapVersionToImage(images []BundleImage) (map[*semver.Version]BundleImage, error) {
```
?